jdaviz/configs/specviz/plugins/viewers.py

```diff
@@ -383,84 +372,79 @@ def add_data(self, data, color=None, alpha=None, **layer_state):
         return result
 
     def _plot_mask(self):
+        if not self.display_mask:
+            return
 
-        _data = self._data[0]
-        _data_trace = self.data_trace_pointer
-
-        if 'mask' in _data.components and self.display_mask:
-
-            # If a mask was already plotted, remove it.
-            if (hasattr(self, 'mask_trace_pointer') and
-                    self.mask_trace_pointer is not None):
-                marks = self.figure.marks[:]
-                self._clean_mask(marks)
-                self.figure.marks = marks
-
-            mask = _data['mask'].data
-
-            # get trace with spectrum.
-            x = self.figure.marks[_data_trace].x
-            y = self.figure.marks[_data_trace].y
-
-            # For plotting markers only for the masked data
-            # points, erase un-masked data from trace.
-            y = np.where(np.asarray(mask) == 0, np.nan, y)
-
-            # there is no 'X' marker option in bqplot
-            self.mask_line_mark = Scatter(scales=self.scales,
-                                          marker='cross',
-                                          x=x,
-                                          y=y,
-                                          stroke_width=0.5,
-                                          # colors=['red'],
-                                          colors=[self._color],
-                                          default_size=25,
-                                          default_opacities=[self._alpha],
-                                          )
-            self.figure.marks = list(self.figure.marks) + [self.mask_line_mark]
-
-            # We added an extra trace. Get pointer to it.
-            self.mask_trace_pointer = len(self.figure.marks) - 1
-
+        # Loop through all active data in the viewer
+        for index, layer_state in enumerate(self.state.layers):
```

Comment:
this is likely catching subsets as well, which is causing a shape-mismatch error when creating a subset and then enabling uncertainties in the plot options.  And probably also won't trigger when new data is added unless the uncertainty marks are all cleared and then recreated.

Reply:
Probably something to do with the `subset_state` attribute, but I don't have time right this moment to look into it. I think that's attached to layers but I'm not 100% sure. Maybe search for that in jdaviz/glue-astronomy.

Reply:
Right! I think that may be the answer, thanks!